File: Skip_Iterator.java
import java.util.HashMap;

class SkipIterator implements Iterator<Integer> {
   Iterator<Integer> nativeIterator;
   Integer nextElement;
   HashMap<Integer, Integer> skipMap;
	public SkipIterator(Iterator<Integer> it) {
	 this.nativeIterator = it;
     this.skipMap = new HashMap<>();
     advance();

    
    }
    void advance() {
        this.nextElement = null;
        while(nativeIterator.hasNext() && nextElement == null) {
            Integer element = nativeIterator.next();
            if(skipMap.containsKey(element)) {
                skipMap.put(element,skipMap.get(element)-1);
                skipMap.remove(element,0);
            } else {
                nextElement = element;
            }
        }
    }

	public boolean hasNext() { //O(1) O(1)
        return nextElement != null;
	}

	public Integer next() { //O(n) O(1)
        Integer result = nextElement;
        advance();
	}

	/**
	* The input parameter is an int, indicating that the next element equals 'val' needs to be skipped.
	* This method can be called multiple times in a row. skip(5), skip(5) means that the next two 5s should be skipped.
	*/ 
	public void skip(int val) { // O(n) O(1)
        if(nextElement != val) {
            skipMap.put(val,skipMap.getOrDefault(val, 0)+1); 
        } else {
            advance();
        }
	}
}

File: Skip_iterator.py
# Time: O(1)
# Space: O(1)


class SkipIterator:

    def __init__(self, iterator):
        #   initialize a global cursor, HashMap to map values to their counts, global iterator
        self.iterator = iterator
        self.skipMap = {}
        self.cursor = None
        self.__seekCursor()

    def hasNext(self) -> bool:

  
        return (self.cursor != None)

    def next(self) -> int:
		
        element = self.cursor
        self.__seekCursor()
        return element

    '''
	The input parameter is an int, indicating that the next element equals 'val' needs to be skipped.
	This method can be called multiple times in a row. skip(5), skip(5) means that the next two 5s should be skipped.
	'''
    def skip(self, val: int) -> None:

        if (self.cursor == val):        #   no need to put count in HashMap if current value
            self.__seekCursor()

        else:                           #   update the count in HashMap
            if (val in self.skipMap):
                self.skipMap[val] += 1
            else:
                self.skipMap[val] = 1

    def __seekCursor(self) -> None:

        #   Time Complexity:    O(n) in the worst case as described above
        #   Space Complexity:   O(n) in the worst case as described above
        self.cursor = None

        nextValue = next(self.iterator, None)             
	# if cursor is not None or if nextValue is None => break the loop
        while(self.cursor == None and  nextValue != None):  
            element = nextValue
	    # if element not in HashMap => assign cursor to that element and break
            if (element not in self.skipMap):               
                self.cursor = element
                break
	  # else update the count, move forward and remove the key if value is 0
            else:                                          
                self.skipMap[element] -= 1
                if (self.skipMap[element] == 0):
                    del self.skipMap[element]

            nextValue = next(self.iterator, None)          


itr = SkipIterator(iter([2, 3, 5, 6, 5, 7, 5, -1, 5, 10]))
print(itr.hasNext())   # true
print(itr.next())      # returns 2
itr.skip(5)
print(itr.next())      # returns 3
print(itr.next())      # returns 6 because 5 should be skipped
print(itr.next())      # returns 5
itr.skip(5)
itr.skip(5)
print(itr.next())      # returns 7
print(itr.next())      # returns -1
print(itr.next())      # returns 10
print(itr.hasNext())   # returns false
print(itr.next())      # None
